src/context/orderReducer.ts

```diff
@@ -0,0 +1,52 @@
+import { OrderCreateDto } from 'dtos/OrderCreateDto';
+import { ProductDto } from 'dtos/ProductDto';
+
+export enum ActionType {
+  Increase = 'ADD_ADDITIONAL_ITEM',
+  Decrease = 'DECREASE_ITEM',
+}
+
+export type Action = {
+  type: ActionType,
+  payload: ProductDto
+}
+
+export const orderReducer = (state: OrderCreateDto, action: Action): OrderCreateDto => {
+  const {type, payload} = action;
+  const existingItem = state.lineItems.find((item) => item.product.id === payload.id);
+  switch (type) {
+  case ActionType.Increase:
+    if (!existingItem) {
+      return { lineItems: [...state.lineItems, { product: payload, quantity: 1, subTotal: payload.price }]};
+    }
+    return {
+      lineItems: [
+        ...state.lineItems.map((item) => {
+          if(item.product.id === payload.id){
+            const subtotal = item.product.price * item.quantity;
+            return {...item, quantity: item.quantity++, subTotal: subtotal };
+          }
+          return item;
+        })
+      ]
+    };
+
+  case ActionType.Decrease:
+    if (existingItem?.quantity === 0) {
+      return {lineItems: [...state.lineItems.filter((item) => item.product.id !== payload.id)]};
+    }
+
+    return {
+      lineItems: [
+        ...state.lineItems.map((item) => {
+          if (item.product.id === payload.id) {
+            return { ...item, quantity: item.quantity-- , subTotal: item.subTotal - item.product.price };
+          }
+          return item;
+        })
+      ]
+    };
+  default:
+    return state;
+  }
+};
```

Comment:
could also create a custom hook that makes using the reducer a bit easier, wouldn't need to pass `state` and `dispatch` as props every time 

something like this maybe? 
```
interface ILocalOrderHook {
  state: OrderCreateDto;
  increase: (product: ProductDto) => void;
  decrease: (product: ProductDto) => void;
}

const useLocalOrder = (): ILocalOrderHook => {
  const [state, dispatch] = useReducer(orderReducer, initialState);

  const increase = (product: ProductDto): void => dispatch({ type: ActionType.Increase, payload: product });
  const decrease = (product: ProductDto): void => dispatch({ type: ActionType.Decrease, payload: product });

  return { state, increase, decrease };
};

```

Reply:
done